### ISSUE TITLE
Fix existing cases of `getFileContents()` usage

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#11166]
after https://github.com/checkstyle/checkstyle/issues/9889

We need to fix all existing cases of `getFileContents()` usage.

cases:
- [ ] MissingOverrideCheck.java
- [ ] PackageAnnotationCheck.java
- [ ] AvoidEscapedUnicodeCharactersCheck.java
- [ ] FallThroughCheck.java
- [ ] PackageDeclarationCheck.java
- [ ] ImportControlCheck.java
- [ ] UnusedImportsCheck.java
- [ ] JavadocMethodCheck.java (1)
- [ ] JavadocStyleCheck.java
- [ ] JavadocTypeCheck.java
- [ ] JavadocVariableCheck.java
- [ ] MissingJavadocMethodCheck.java
- [ ] MissingJavadocPackageCheck.java
- [ ] MissingJavadocTypeCheck.java
- [ ] WriteTagCheck.java
- [ ] OuterTypeFilenameCheck.java
- [ ] RegexpCheck.java
- [ ] RegexpSinglelineJavaCheck.java
- [ ] MethodLengthCheck.java
- [ ] EmptyLineSeparatorCheck.java
- [ ] JavadocMetadataScraper.java

it is better to resolve each case in separate PR. Each PR should remove the suppression to show the issue is fixed. https://github.com/haanhvu/checkstyle/commit/14391708cba2675bb2cd56135533bc82d6c120c2#diff-c1ffc75b7152435a4d5d40d7ee9b32b57402959b1ad471a41e753ea67554378b

**Comment by remal** [comment]
@romani I have some custom checks for JavaDocs and, I'm afraid, it's not quite clear what exactly is supposed to be used instead of `getFileContents()`. Could you (or someone else) explain what is supposed to be used?

My goal is it check that JavaDoc is presented on every element that matches specific condition, has some tag (`@author`) and the tag's value is equal to some constant.

**Comment by strkkk** [comment]
@remal instead of `getContents()` checks should use only AST and do not rely on text in file.
About the case you described - it seems to me that you dont need a custom check for it, MatchXPathCheck can provide same functionality.

**Comment by romani** [comment]
@remal , please look at
https://github.com/checkstyle/checkstyle/blob/0b3a2606e535eb14d2d9dcd6c79d00db17c22f91/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AtclauseOrderCheck.java#L152

With such inheritance, you will get whole ast of javadoc content.
The only downside is performance degradation, but you can measure how critical it for you in your projects.

**Comment by romani** [comment]
Another approach is to override
https://github.com/checkstyle/checkstyle/blob/0b3a2606e535eb14d2d9dcd6c79d00db17c22f91/src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java#L85 to return "true" and your Check will get AST with all comment as plain text and you can parsing/matching content of comments as you need. You can find a lot of examples in our code of such override.
Such approach does not have any performance problems.

**Comment by romani** [comment]
Another approach is to override
https://github.com/checkstyle/checkstyle/blob/0b3a2606e535eb14d2d9dcd6c79d00db17c22f91/src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java#L85 to return "true" and your Check will get AST with all comment as plain text and you can parsing/matching content of comments as you need. You can find a lot of examples in our code of such override.
Such approach does not have any performance problems.

**Comment by romani** [comment]
I pretty sure we need individual issue for each Check, updates are not simple and there might be breaking changes or kind of new functionality

**Comment by strkkk** [comment]
@Kevin222004 I would say it is better to pick up some other issue (with gsoc label), this changed are not easy for new contributors.

**Comment by kenyee** [comment]
For our usage of getFileContents(), we just care about the filename.  Is there a better way to get that?

**Comment by romani** [comment]
Please use
https://github.com/checkstyle/checkstyle/blob/b09a9ce9d4f48c179f3d5647b6656e5737462656/src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java#L306

**Comment by aayushRedHat** [comment]
I am on JavadocStyleCheck

**Comment by romani** [comment]
@aayushRedHat , please  be aware that refactoring of javadoc check needs changing extension of abstract class to get AST of javadoc. It might be bigger update than for non javadoc Checks. Do not hesitate to set draft PR to share with us your direction and double check that you care right path.

**Comment by Kevin222004** [comment]
I am on JavadocMethodCheck

**Comment by romani** [comment]
@Kevin222004 , please do not deal with Javadoc* Checks, it will require big redesign we do not have plan for this, please deal with regular Checks for now.

**Comment by mahfouz72** [comment]
@romani I want to work on this. I can take EmptyLineSeparatorCheck. Is there anything I should keep in mind before starting? and
is it ok to split work for this Check in separate PRs to ease work and review because there are two usages of `getFileContents()` in this check or full migration should be done in one PR?

**Comment by romani** [comment]
@mahfouz72, please help us to refactor this Check, yes, few PRs are good to ease review.

**Comment by mohitsatr** [comment]
i think i can help with the remaining cases

**Comment by mohitsatr** [comment]
https://github.com/checkstyle/checkstyle/blob/10af01481ac3ed056c4cc4f1085a4e4f043c243c/src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java#L295-L296

can we just make an simple api like in AbstractCheck, something like this:
```
getSingleLineComments(){
 return context.get().fileContents.getSinglelinecomments()
}
```

like the way, we are currently accessing file paths with `getFilePath()` in AbstractCheck?

**Comment by romani** [comment]
No, all should migrate to AST, there might be some exceptional usage of content, but  it will be well explained cases.

**Comment by mohitsatr** [comment]
private String unitAbbrev2 = "\u03bcs"; // violation

     private String unitAbbrev3 = "\u03bcs"; // violation 


      |--VARIABLE_DEF -> VARIABLE_DEF [21:8]
        |   |--MODIFIERS -> MODIFIERS [21:8]
        |   |   |--SINGLE_LINE_COMMENT -> // [19:48]
        |   |   |   `--COMMENT_CONTENT ->  violation\n [19:50]
        |   |   `--LITERAL_PRIVATE -> private [21:8]
        |   |--TYPE -> TYPE [21:16]
        |   |   `--IDENT -> String [21:16]
        |   |--IDENT -> unitAbbrev4 [21:23]
        |   |--ASSIGN -> = [21:35]
        |   |   `--EXPR -> EXPR [21:37]
        |   |       `--STRING_LITERAL -> "\u03bcs" [21:37]
        |   `--SEMI -> ; [21:46]
        |--METHOD_DEF -> METHOD_DEF [23:8]
        |   |--MODIFIERS -> MODIFIERS [23:8]
        |   |   |--SINGLE_LINE_COMMENT -> // [21:48]
        |   |   |   `--COMMENT_CONTENT ->  violation\n [21:50]
        |   |   `--LITERAL_PUBLIC -> public [23:8]
        |   |--TYPE -> TYPE [23:15]
        |   |   `--IDENT -> Object [23:15]

while brainstorming for this problem, i noticed comments aren't at right places, then I found that this weird comment placement in the ast has been pointed out several times before. but no one's attempted to fix it yet. according to me, it makes more sense to fix #6233 first. 

it looks like it is due to the way `appendHiddenCommentNodes` works.

is there any reason we have not fixed it yet? if no, I'm willing to work on it.

**Comment by romani** [comment]
If you see approved issue as not fixed, reason is only that we have no time/people to fix it. Please be welcome, but careful as that issue is not simple.

**Comment by roulpriya** [comment]
These files no longer contain reference to `getFileContents()`
`ImportControlCheck.java`
`FallThroughCheck.java`

**Comment by Rohanraj123** [comment]
Hey @romani I tried using AST based methods for WriteTagCheck.java. But I found that In checkstyle's AST comments are not  direct siblings of class/method nodes.

**Comment by Rohanraj123** [comment]
Not necessary, I think just overriding the `isCommentNodeRequired()` func of AbstractCheck and making it return true will work in enabling the comments as a AST sibling. 
```

 @Override
    public boolean isCommentNodesRequired() {
        return true;
    }
```

@romani

**Comment by Hariom-kr** [comment]
Hii, @romani,  How should I replace getFileContents(). in UnusedImportsCheck? What's the correct AST-based approach ?

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle/blob/a3223634c34b592713d3fa9982b53e5f02f1d2a5/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java#L317-L320

Attention to this part. Check just needs javadoc.
https://checkstyle.org/writingchecks.html#Comment_Tokens_in_AST

Look around in other checks how `isCommentNodesRequired()` help Checks to get content of javadoc.

**Comment by ivamly** [comment]
Hello, @romani! I'd like to contribute by working on [JavadocMethodCheck.java](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java). Is this available for me to take?

**Comment by romani** [comment]
@ivamly, yes, all not fixed are available, I do not see PR also .
But be aware that it might need usage of javadoc AST, please look around how others javadoc checks are done. If any questions, please create draft PR and we will discuss in PR

**Comment by Hariom-kr** [comment]
I  am On AvoidEscapedUnicodeCharacters.java

**Comment by romani** [comment]
@ivamly, how you are doing?
If you are stuck, let me know, you can take some simple Check. JavadocMethod is very complicated, better to do it after gaining more experience.

I don't see this JavadocMethod in list of checkboxes, it might be omitted on purpose, as it is not simple, and needs very special attention.

**Comment by ivamly** [comment]
Hello @romani! Thank you for your message and the guidance.

I'm not confident I can tackle the JavadocMethodCheck properly at the moment, so I think it would be best for me to take on a different task.

Could you please recommend a specific check from the available ones? Alternatively, should I just pick any unassigned task that is not Javadoc-related?

**Comment by romani** [comment]
Try any unassigned.
Try FallThroughCheck.

If you ever again see that task is too complicated, pin me, we can find something else or find a easier way of doing it.

Thanks a lot for your help in our project.

**Comment by ivamly** [comment]
@romani By the way, the main issue description seems to be outdated.

It lists the following classes as requiring fixes:

- `FallThroughCheck.java`
- `ImportControlCheck.java`
- `WriteTagCheck.java`

However, judging by the [suppression.xml](https://github.com/checkstyle/checkstyle/blob/master/config/suppressions.xml) file and the code, it appears these have already been fixed.

**Comment by romani** [comment]
Thanks a lot!!
Updated.

Please help to migrate UnusedImportsCheck

**Comment by ivamly** [comment]
I'm not quite sure what specific help is needed - I can see there's already a [PR](https://github.com/checkstyle/checkstyle/pull/18059) in progress that seems to be addressing all the issues with UnusedImportsCheck.

**Comment by romani** [comment]
Thanks a lot for comment.
Yes, looks like this issue is about to be finished.

Last 2 PRs are are close to completion

**Comment by Aman-Baliyan** [comment]
i am on UnusedImportsCheck.java

**Comment by Tasneemmohammed0** [comment]
I'm on `UnusedImportsCheck`.

**Comment by Tasneemmohammed0** [comment]
I'm on `JavadocVariableCheck`.

**Comment by romani** [comment]
> For javadoc please consult with maintainers before starting

moved to separate issues:
https://github.com/checkstyle/checkstyle/issues/19144
https://github.com/checkstyle/checkstyle/issues/19145
https://github.com/checkstyle/checkstyle/issues/19146
https://github.com/checkstyle/checkstyle/issues/19147
https://github.com/checkstyle/checkstyle/issues/19148
https://github.com/checkstyle/checkstyle/issues/19149
https://github.com/checkstyle/checkstyle/issues/19150

**Comment by romani** [comment]
There are already 2 PRs for non javadoc Check, so this issue is almost closed, we just need to merge already create PRs.

Plase do  not send PRs for this issue !!!!! use other [good fifth issues](https://github.com/checkstyle/checkstyle/issues?q=state%3Aopen%20label%3A%22good%20fifth%20issue%22)

**Comment by romani** [comment]
All changes are merged

**Comment by romani** [comment]
Reopened to allow us change reason of suppression in javadoc Checks
